Add ntp migration actors

mlichvar opened the pull request:
Add CheckNtp actor to check if the ntpd and/or ntpdate services are
enabled, active and their configuration files exist. Also, save the
configuration files in a string containing a base64-encoded compressed
tarball to avoid having to deal with .rpmsave files when the packages
are removed in the rpm upgrade.

Add MigrateNtp actor to enable the chronyd service, restore the
ntp/ntpdate configuration files, and run the included ntp2chrony
migration script. Produce a warning with a suggestion to check the
generated /etc/chrony.conf if some lines in ntp.conf were ignored in the
migration.

pirat89 commented:
leapp-ci build

vojtechsokol commented:
leapp-ci build

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/actor.py:
```diff
@@ -0,0 +1,76 @@
+                    f.write('driftfile /var/lib/ntp/drift\n')
+                    f.write('restrict default ignore nomodify notrap nopeer noquery\n')
+
+            if 'ntpdate' in decision.migrate_services:
```
nit: oneliner will be easier to read

step_tickers = '/etc/ntp/step-tickers' if 'ntpdate' in decision.migrate_services else ''

mlichvar replied:
Ok. Fixed.

fernflower commented on repos/system_upgrade/el7toel8/actors/checkntp/actor.py:
```diff
@@ -0,0 +1,74 @@
+        active = False
+        enabled = False
+        try:
+            ret = call(['systemctl', 'is-active', name])
```
unused var?

mlichvar replied:
Fixed.

fernflower commented on repos/system_upgrade/el7toel8/actors/checkntp/actor.py:
```diff
@@ -0,0 +1,74 @@
+            self.log.debug('{} is not active'.format(name))
+
+        try:
+            ret = call(['systemctl', 'is-enabled', name])
```
Can 2 check procedures be merged into one? Or reuse code?

mlichvar replied:
This was simplified in the latest commit.

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/actor.py:
```diff
@@ -0,0 +1,76 @@
+                # Use a minimal secure ntp.conf if ntpd was disabled
+                ntp_conf = '/etc/ntp.conf.nosources'
+                with open(ntp_conf, 'w') as f:
+                    f.write('# This file was created to migrate ntpdate configuration to chrony\n')
```
Maybe move minimal config definiton contents to some variable instead of writing it here line per line?

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/actor.py:
```diff
@@ -0,0 +1,76 @@
+
+            # Report ignored lines in ntp.conf, except 'disable monitor' from
+            # the default ntp.conf
+            for ignored_line in ntp_configuration.ignored_lines:
```
Can't get the logic here. IIUC the first line that is ignored but isn't 'disable_monitor' will break the loop (not processing any other lines) and go straight to L72, while L66 will be reached if and only if ntp_configuration.ignored_lines == ['disable monitor'].
Is this intended? In that case - can you please rewrite it to make it more obvious?

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,664 @@
+                    .format(self.root_dir, service)):
+                self.enabled_services.add(service)
+        if self.verbose > 0:
+            print("Enabled services found in /etc/systemd/system: " +
```
please you string formatting

pirat89 replied:
In this case, we are talking about script that is used probably somewhere else as well and tested already. I would keep it as it is as it can follow even different practices in other projects. I think this is something what we do not have mentioned anywhere even best practices - how we will look at the third party code. Of course, we can keep it as it is code going just into the project, but... I feel that we should think that yet.

mlichvar replied:
Should be fixed in HEAD.

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,664 @@
+        path = self.root_dir + self.step_tickers_path
+        if not os.path.isfile(path):
+            if self.verbose > 0:
+                print("Missing " + path)
```
Use logging instead of prints?

mlichvar replied:
Makes sense. Should be fixed in HEAD.

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,664 @@
+                print("Missing " + path)
+            return
+
+        with open(path, encoding="latin-1") as f:
```
Is this the default encoding for every single case? Imho worth adding an encoding param to the fucntion and not having magic constants

mlichvar replied:
We don't know the encoding, so files are read and written using the same (8-bit transparent) encoding. I've added a variable for the encoding to make it more clear.

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,664 @@
+                self.step_tickers.extend(words)
+
+    def parse_ntp_conf(self, path=None):
+        if path is None:
```
nit: path = self.root_dir + self.ntp_conf_path if not path else path

mlichvar replied:
This looks like a matter of preference.

fernflower replied:
Please use os.path.join for all path building operations

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,664 @@
+        if words:
+            if name in ["server", "peer", "pool"]:
+                return self.parse_source(name, words)
+            elif name == "fudge":
```
nit: elif -> if for all cases with return

Actually you can even have an action map like
action_map = {"tos": self.parse_tos,
                      "includefile": self.parse_includefile}
and call the majority of methods as action_map[key](words)

But it's just a suggestion of course

mlichvar replied:
That would work, but it would need to be applied to the whole parser and I'm not sure if it's worth the trouble. I prefer consistency.

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,664 @@
+            elif tos[0] == "orphan":
+                orphan_stratum = int(tos[1])
+            else:
+                assert False
```
assert False?

mlichvar replied:
That would fail is there was a unexpected tos (i.e. bug). Anyway, I've reworked this for the other issue with int() failing.

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,664 @@
+            address = source["address"]
+            if address.startswith("127.127.1."):
+                if address in self.fudges and "stratum" in self.fudges[address]:
+                    local_stratum = int(self.fudges[address]["stratum"])
```
Can this raise a ValueError during convertion?

mlichvar replied:
Yes, good point. It should be now fixed.

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,664 @@
+
+        conf += self.get_chrony_conf_sources()
+
+        conf += "# Record the rate at which the system clock gains/losses time.\n"
```
Again imho worth considering having the base of config in some variable.

mlichvar replied:
There are many strings in the code like this. I don't think it makes sense to put only some in variables.

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,664 @@
+            os.rename(path, path + ".old")
+
+        with open(os.open(path, os.O_CREAT | os.O_WRONLY | os.O_EXCL, mode), "w",
+                  encoding="latin-1") as f:
```
please do not hardcode encoding

mlichvar replied:
Same as above.

pirat89 commented on repos/system_upgrade/el7toel8/actors/migratentp/actor.py:
```diff
@@ -0,0 +1,76 @@
+
+            # Unpack archive with configuration files
+            #call(['/bin/sh', '-c', 'base64 -d | tar xv'], stdin=decision.config_tgz64)
+            p = Popen(['/bin/sh', '-c', 'base64 -d | tar xz -C /'], stdin=PIPE, stdout=None, stderr=None)
```
anyway, use of the `subprocess` module and Popen especially is something we do not want to have in the code in future. But agree with @fernflower, python modules

mlichvar replied:
Ok. Rewritten to use the tarfile, io, base64 modules.

fernflower commented:
leapp-ci build

vojtechsokol commented:
leapp-ci build

vojtechsokol commented:
Tests are failing because of missing import: `ImportError: No module named ipaddress`.
Status is not updated to `failure` due to a bug in our CI.

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,664 @@
+            self.ignored_directives.add(name)
+            return False
+
+        return True
```
Maybe convert 'return True\False' to 'return a_dict_of_parsed_values'? This will make it instantly more obvious which exactly params were processed and make writing tests for actual conversion a bit easier, as this will allow section-by-section convertion testing

mlichvar replied:
How would you distinguish between an error in the parsing and nothing parsed?

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,664 @@
+            try:
+                version = ipaddress.ip_address(source["address"]).version
+                if (ipv4_only and version != 4) or (ipv6_only and version != 6):
+                    return False
```
same thing here, please see comment for changing return False to return a_dict_of_processed_values

fernflower replied:
Is it a bug in time source parsing with include file https://github.com/mlichvar/ntp2chrony/blob/master/tests/data/chronyconfs/3_chrony.conf#L4? Several lines appear twice in generated config.

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,664 @@
+        self.ignored_lines = []
+
+        #self.detect_enabled_services()
+        self.parse_step_tickers()
```
I'd move calls to logic outside of __init__, intoduce addional parse() class-method and call it specifically from main(). 
I also feel that some methods can be made _private.

mlichvar replied:
That would mean parsing could be performed twice or not at all? That doesn't make much sense to me. The class represents a single instance of ntp configuration.

pirat89 commented:
@mlichvar Do not forget to rebase branch with master with new update as the recent changes has to be applied to be able to upgrade.

fernflower commented:
leapp-ci build

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,670 @@
+
+            prefix = "ASCII" if len(password) <= 20 else "HEX"
+
+            for first, last in self.trusted_keys:
```
iiuc L600-L605 are the same as
trusted = any(id in range(first, last+1) for (first, last) in self.trusted_keys)

mlichvar replied:
Similar, but faster for wide ranges and maybe slightly more readable.

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,670 @@
+        keys += "\n"
+
+        for key in self.keys:
+            id = key[0]
```
We don't have it in the guidelines yet but please don't use builtins (id, type) as vars

key_id, key_type, password = key

mlichvar replied:
Makes sense. I'll fix that.

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,670 @@
+        with open(os.open(path, os.O_CREAT | os.O_WRONLY | os.O_EXCL, mode), "w",
+                  encoding=self.file_encoding) as f:
+            logging.info("Writing %s", path)
+            f.write(u"" + content)
```
f.write(unicode(content))?

mlichvar replied:
Is there a difference?

fernflower replied:
As in unnecessary string concatenation with forceful unicode convertion vs direct unicode convertion

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,670 @@
+                if pool[0][1]["options"] != pool[i][1]["options"]:
+                    break
+            else:
+                assert len(pool) > 2
```
Unfortunately asserts are against leapp guidelines https://github.com/oamg/leapp-guidelines/blob/master/python-coding-guidelines.md

mlichvar replied:
The guideline just says asserts shouldn't mutate (`len(pool) > 2` certainly doesn't).

bocekm commented:
@mlichvar, we are changing license of this repository from AGPL 3.0 to Apache 2.0 under PR https://github.com/oamg/leapp-repository/pull/112. Is it ok with you?

mlichvar commented:
> @mlichvar, we are changing license of this repository from AGPL 3.0 to Apache 2.0 under PR #112. Is it ok with you?

Fine with me. I guess I need to relicense the ntp2chrony script?

pirat89 commented:
@mlichvar Regarding the [Fedora Licensing](https://fedoraproject.org/wiki/Licensing:Main?rd=Licensing) page, ASL 2.0 is not compatible with GPLv2.

bocekm commented:
@mlichvar, yes please. Apache 2.0 is not compatible with GPL 2.0. Either change the license to a compatible one or remove the license notice completely if you agree with Apache 2.0 for the script.

I think you can also remove the `#!/usr/bin/python` and `if __name__ == "__main__":
    main()`.

mlichvar commented:
Relicensing only the leapp copy would mean it couldn't get any later improvements from upstream (made by other people). I think I'll change the license to MIT, so it stays compatible with GPLv2.

bocekm commented:
And where's the upstream copy of the script?

mlichvar commented:
It's here: https://github.com/mlichvar/ntp2chrony/blob/master/ntp2chrony.py

bocekm commented:
Ok, yeah. That will work. Changing it to MIT on both places, here and in the upstream.
By the way, in the upstream [you changed](https://github.com/mlichvar/ntp2chrony/commit/dcf7a05070ebdf7bbe1d59f3a53c68e8795c3d5f) the license to [2-Clause BSD](https://opensource.org/licenses/BSD-2-Clause), not [MIT](https://opensource.org/licenses/MIT).

mlichvar commented:
Oops. I'll fix that. Thanks for checking.

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,676 @@
+                continue
+            if not source["address"].endswith(".pool.ntp.org"):
+                continue
+            m = re.match("^([0123])(\\.\\w+)?\\.pool\\.ntp\\.org$", source["address"])
```
This will add '.' to the zone name, is it intended?

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,676 @@
+        convert_servers = set()
+        for zone, pool in pools.items():
+            pool.sort()
+            if [number for number, source in pool] != ["0", "1", "2", "3"]:
```
If there are not exactly 4 servers (0, 1, 2, 3) the check on L376-L380 won't happen. Is it intended?

fernflower replied:
L370-L396 look overcomplicated. 
Both convert_ and remove_ sets are unnecessary, you don't need convert_set and the cycle L388-L394 at all as you can change the object in place. len(pool) is 4 all the time, this cycle L376-L380 can also be substituted for an easily readable remove_set update.

mlichvar replied:
Should be fixed now.

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,676 @@
+        remove_servers = set()
+        convert_servers = set()
+        for zone, pool in pools.items():
+            pool.sort()
```
Do you mean lexicographic check here, or are you missing int conversion for pool[0]?

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,676 @@
+        conf = ""
+        for network in sorted(allowed_networks, key=lambda n: (n.version, n)):
+            ip_versions.add(network.version)
+            if network.num_addresses > 1:
```
nit: conf += "cmdallow {}\n".format(network if network.num_addresses > 1 else network.network_address)

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,676 @@
+        for source in self.time_sources:
+            if source["type"] != "server":
+                continue
+            if not source["address"].endswith(".pool.ntp.org"):
```
Unnecessary check, will be covered by regex at L361

mlichvar commented on repos/system_upgrade/el7toel8/actors/migratentp/actor.py:
```diff
@@ -0,0 +1,92 @@
+    tags = (ApplicationsPhaseTag, IPUWorkflowTag)
+
+    def extract_tgz64(self, s):
+        import base64, io, tarfile
```
Fixed.

mlichvar commented on repos/system_upgrade/el7toel8/actors/checkntp/actor.py:
```diff
@@ -0,0 +1,82 @@
+
+    def get_tgz64(self, filenames):
+        # Return a base64-encoded gzipped tarball
+        import base64, io, tarfile
```
Fixed.

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,669 @@
+        return True
+
+    def parse_includefile(self, words):
+        path = self.root_dir + words[0]
```
Please use os.path.join for all path building operations

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,669 @@
+        if not self.step_tickers_path:
+            return
+
+        path = self.root_dir + self.step_tickers_path
```
Please use os.path.join for all path building operations

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,669 @@
+
+    def parse_keys(self, words):
+        keyfile = words[0]
+        path = self.root_dir + keyfile
```
Please use os.path.join for all path building operations

mlichvar replied:
Fixed.

fernflower commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,669 @@
+            if address.startswith("127.127."):
+                if address.startswith("127.127.1."):
+                    continue
+                assert False
```
Sorry, not getting it - what is this assert used for? What special case are you addressing by it?

shaded-enmity replied:
+1, what would make more sense to me:
```
if not address.startswith('127.127.1.'):
   raise Exception('Some meaningful message as to why this is not correct')
```

mlichvar replied:
asserts catch bugs in the code. In a bug-free code they do nothing. In this case there are two places in the code that need to be consistent with each other. The parser and the config generator. If the parser was modified to handle another refclock, but the config generator wasn't, the assert would trigger.

shaded-enmity replied:
@mlichvar This is handling user provided data, there may not be a bug in the code, but the data can be ill-formed, and you NEVER use asserts for that. Please also read the guidance about "In-your-face error handling", the point here is that these things MAY bubble up to users, and hence the very clear distinction between:
```
AssertionError: False
```
And
```
NtpMigrationError: Configuration in /some/path contains multiple refclocks
```

mlichvar replied:
But that doesn't apply to this case. No user provided data can trigger that assert. The parser ignores all refclocks except 127.127.1.0, so they are listed in the generated file as ignored directives. The assert will trigger only if the parser is updated to handle other refclocks, but the generator is not, i.e. there is a bug in the logic of the code.

shaded-enmity replied:
Thanks for the clarification @mlichvar - in this case I'd suggest removing the two checks altogether and flesh them out into a unit test, because that's what it is, we will not be testing if someone made an incompatible change to he parser when running on a production system, we need to test for that before we release.

As a side note, about asserts, since I'm the one who wrote that section of the guidelines, the "mutating assert" scenario is spelled out merely as a worst case scenario, but the more fundamental takeaway is that relying on assert is simply making assumptions as to how people are executing our code. We cannot make these assumptions.

mlichvar replied:
Right, there should be tests with non-LOCAL refclocks. And those may or may not trigger the assert depending on what bug the code actually has. Checking the output is important, but it may not catch everything.

shaded-enmity commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,669 @@
+                    elif option[0] == "true":
+                        conf += " trust"
+                    else:
+                        assert False
```
Ditto here, please raise an exception with a meaningful message.

mlichvar replied:
I don't think a different exception would be appropriate here. I have added a comment to make it more clear what the assertion is supposed to catch.

shaded-enmity replied:
We try to provide all the necessary context to the users to fix the actual problem, why am I supposed to read code when the actual problem is in my data?

mlichvar replied:
Users will not see this. Please have a look at the actual code.

mlichvar commented:
Added unit tests.

mlichvar commented on repos/system_upgrade/el7toel8/actors/migratentp/tests/unit_test.py:
```diff
@@ -0,0 +1,95 @@
+from leapp.libraries.actor import library
+from leapp.libraries.common import reporting
+
+class report_generic_mocked(object):
```
I can rename the classes, but I've copied this from other tests that are already included in leapp-repository, so it would be nice to have some consistency here.

mlichvar commented on repos/system_upgrade/el7toel8/actors/migratentp/tests/unit_test.py:
```diff
@@ -0,0 +1,95 @@
+    def __init__(self):
+        self.called = 0
+
+    def __call__(self, **report_fields):
```
No, I'm not familiar with the mock library. I've just followed the leapp recommendations for writing tests.

shaded-enmity commented on repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py:
```diff
@@ -0,0 +1,671 @@
+                if address.startswith("127.127.1."):
+                    continue
+                # No other refclocks are expected from the parser
+                assert False
```
@mlichvar I can still see the `assert` here (and below) and want to reiterate that the conditions that make the code reach such code paths should be rather handled by proper unit test, if we get here, it's a code error and we need to catch that early. If you need help with the tests, let me know.

mlichvar replied:
There are tests that would trigger both asserts in the upstream repository:
https://github.com/mlichvar/ntp2chrony/blob/master/tests/data/ntpconfs/6_ntp.conf
https://github.com/mlichvar/ntp2chrony/blob/master/tests/data/ntpconfs/7_ntp.conf

mlichvar commented:
It looks like the code is catching a wrong CalledProcessError, after leapp has switched to its own? Should I fix the first commit in the series?

vinzenz commented:
@mlichvar You are right that's the reason - You can also just add another patch on top of it, as we have to squash this PR anyway.
And please when you are on it - Transform please the description fields of the Actors to class docstrings instead and try to avoid using double quotes for non docstrings.

Thanks.

fernflower commented:
Yeah, that's the issue. Good news is that after that is fixed the remaining part seems to be ok.
I used a different approach, by switching to a quiet mode with checked=False you implemented, but either way it is ok.
```
def check_service(name):
    for state in ['active', 'enabled']:
        res = run(['systemctl', 'is-{}'.format(state), name], checked=False)
        res_state = res['stdout'].strip()
        api.current_logger().debug('{} is {}'.format(name, res_state))
        if res_state != state:
            return False

    return True
```
You can add a commit on top or fixup the change into first commit - it doesn't matter imho as I suppose all commits will be squashed upon merge.

mlichvar commented:
Ok, I fixed the imports to use the right exception (with checked=False the function would always return True) and fixed the actor descriptions to use docstrings. If you want to convert ntp2chrony.py from quotes to double quote, please feel free.

Thanks for your help with this PR.

mlichvar commented on repos/system_upgrade/el7toel8/actors/checkntp/libraries/library.py:
```diff
@@ -0,0 +1,70 @@
+
+
+# Check if a file exists
+def is_file(name):
```
The function is mocked in the unit tests.
